### ISSUE TITLE
define-generic and define-method creates global bindings in scope

### ISSUE THREAD

**Issue by ktakashi** [ktakashi/sagittarius-scheme#198]
**[Original report](https://bitbucket.org/ktakashi/sagittarius-scheme/issues/193) by Takashi Kato (Bitbucket: [ktakashi](https://bitbucket.org/ktakashi), GitHub: [ktakashi](https://github.com/ktakashi)).**

----------------------------------------

This should be `&undefined`
```
#!scheme
(import (clos user))

(let ()
  (define-generic foo))
foo
```

**Comment by ktakashi** [comment]
**Original comment by Takashi Kato (Bitbucket: [ktakashi](https://bitbucket.org/ktakashi), GitHub: [ktakashi](https://github.com/ktakashi)).**

----------------------------------------

The culprit is `%ensure-generic-function` procedure. The procedure creates a generic function in passed library.
This is needed to make generic function implicitly (and unfortunately, there's a library depending on this behviour).

What we can do without removing implicit generation is that adding macro environment check. If the environment is not toplevel, then we don't have to make implicitly.

**Comment by ktakashi** [comment]
**Original comment by Takashi Kato (Bitbucket: [ktakashi](https://bitbucket.org/ktakashi), GitHub: [ktakashi](https://github.com/ktakashi)).**

----------------------------------------

Adding local variable check for generic functions (Fixes #193)
